UnityProject/Assets/Scripts/Player/PlayerCrafting.cs

```diff
@@ -0,0 +1,350 @@
+		/// <param name="possibleIngredients">
+		/// 	The ingredients(or reagent containers) that may be used for crafting.
+		/// </param>
+		/// <param name="possibleTools">The tools that may be used for crafting.</param>
+		/// <returns>True if a crafting action has started, false otherwise.</returns>
+		public bool TryToStartCrafting(
+			CraftingRecipe recipe,
+			List<CraftingIngredient> possibleIngredients,
+			List<ItemAttributesV2> possibleTools
+		)
+		{
+			if (!CanCraft(recipe, possibleIngredients, possibleTools))
+			{
+				return false;
+			}
+
+			StartCrafting(recipe, possibleIngredients, possibleTools);
+			return true;
+		}
+
+		/// <summary>
+		/// 	Unsafely starts a new crafting action even if the recipe's requirements were not fulfilled.
+		/// </summary>
+		/// <param name="recipe">The recipe that the player is trying to craft to.</param>
+		/// <param name="possibleIngredients">
+		/// 	The ingredients(or reagent containers) that may be used for crafting.
+		/// </param>
+		/// <param name="possibleTools">The tools that may be used for crafting.</param>
+		private void StartCrafting(
+			CraftingRecipe recipe,
+			List<CraftingIngredient> possibleIngredients,
+			List<ItemAttributesV2> possibleTools
+		)
+		{
+			if (recipe.CraftingTime >= 1) {
+				Chat.AddExamineMsgFromServer(
+					playerScript.gameObject,
+					$"You are trying to craft {recipe.RecipeName}..."
+				);
+			}
+
+			if (recipe.CraftingTime <= float.Epsilon)
+			{
+				FinishCrafting(recipe, possibleIngredients, possibleTools);
+				return;
+			}
+			StandardProgressAction.Create(
+				craftProgressActionConfig,
+				() => TryToFinishCrafting(recipe)
+			).ServerStartProgress(playerScript.registerTile, recipe.CraftingTime, playerScript.gameObject);
+		}
+
+		/// <summary>
+		/// 	Tries to finish a crafting action.
+		/// 	May use all reachable items from a tile that a player is directed to.
+		/// 	May use all tools that a player holds in his hands.
+		/// </summary>
+		/// <param name="recipe">The recipe to try to craft.</param>
+		/// <returns>True if we can spawn the recipe's result, false otherwise.</returns>
+		public bool TryToFinishCrafting(CraftingRecipe recipe)
+		{
+			return TryToFinishCrafting(recipe, GetPossibleIngredients(), GetPossibleTools());
+		}
+
+		/// <summary>
+		/// 	Tries to finish a crafting action.
+		/// </summary>
+		/// <param name="recipe">The recipe to try to craft.</param>
+		/// <param name="possibleIngredients">
+		/// 	The ingredients(or reagent containers) that may be used for crafting.
+		/// </param>
+		/// <param name="possibleTools">The tools that may be used for crafting.</param>
+		/// <returns>True if we can spawn the recipe's result, false otherwise.</returns>
+		public bool TryToFinishCrafting(
+			CraftingRecipe recipe,
+			List<CraftingIngredient> possibleIngredients,
+			List<ItemAttributesV2> possibleTools
+		)
+		{
+			if (!CanCraft(recipe, possibleIngredients, possibleTools))
```

Comment:
please change all usages of ``!`` for ``== false``. I also hate it, but it is the code style we use.